lib/Sema/SemaCast.cpp

```diff
@@ -2565,7 +2565,15 @@ void CastOperation::CheckCStyleCast() {
     unsigned TypeKind = 0;
     bool HasUncheckedType = DestType->hasUncheckedType(TypeKind);
     bool HasVariadicType = DestType->hasVariadicType();
-    if (HasUncheckedType || HasVariadicType) {
+    bool ConstructsNullPointer = false;
+    if (HasUncheckedType)
+      if (DestType->isVoidPointerType() &&
+          !SrcExpr.isInvalid()) {
+        const IntegerLiteral *Lit = dyn_cast<IntegerLiteral>(SrcExpr.get());
+        if (Lit && !Lit->getValue())
+         ConstructsNullPointer = true;
+      }
```

Comment:
I think it might be clearer if you use the following method on SrcExpr, which will do the check better than we ever can: https://clang.llvm.org/doxygen/classclang_1_1Expr.html#aa1d79a39abcf576a82b97cff17f893f8

This will also do the right thing for other kinds of NULL (yes, there are other null expressions that clang supports).

As for `(int*)0`, it seems reasonable not to support it yet. If someone has written their own `NULL_INT_PTR` macro in a header, then they have to change it if they want to use it in a checked pointer.

`(ptr<int>)0` is allowed as this is entirely checked.